File: python/leetcode/413.py
def numberOfArithmeticSlices(self, A):
    dp, i = [0,0], 1
    for j in xrange(2,len(A)):
        if A[j]-A[j-1] == A[j-1]-A[j-2]:
            dp.append(opt[j-1]+i)
            i += 1
        else:
            dp.append(opt[j-1])
            i = 1
    return dp[-1]

File: cpp/leetcode/413.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        int n = A.size();
        if (n < 3) return 0;
        vector<int> dp(n, 0); // dp[i] means the number of arithmetic slices ending with A[i]
        if (A[2] - A[1] == A[1] - A[0]) dp[2] = 1; 
        int res = dp[2];
        for (int i = 3; i < n; ++i) {
            if (A[i] - A[i-1] == A[i-1] - A[i-2]) 
                dp[i] = dp[i-1] + 1;
            res += dp[i]; 
        }
        return res;
    }
};